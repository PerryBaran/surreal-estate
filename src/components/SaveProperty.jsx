import React from "react";
import PropTypes from "prop-types";
import style from "../styles/saveProperty.module.css";

const SaveProperty = ({
  userId,
  _id,
  favouriteId,
  onSaveProperty,
  onRemoveProperty,
}) => {
  if (userId) {
    if (favouriteId) {
      return (
        <button
          type="button"
          onClick={() => onRemoveProperty(favouriteId)}
          className={style["remove-property"]}
        >
          <span>Saved</span>
        </button>
      );
    }
    return (
      <button
        type="button"
        onClick={() => onSaveProperty(_id)}
        className={style["save-property"]}
      >
        Save
      </button>
    );
  }
  return null;
};

SaveProperty.defaultProps = {
  favouriteId: null,
};

SaveProperty.propTypes = {
  _id: PropTypes.string.isRequired,
  userId: PropTypes.string.isRequired,
  favouriteId: PropTypes.string,
  onSaveProperty: PropTypes.func.isRequired,
  onRemoveProperty: PropTypes.func.isRequired,
};

export default SaveProperty;
