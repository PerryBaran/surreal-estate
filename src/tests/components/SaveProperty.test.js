import React from "react";
import { render, screen } from "@testing-library/react";
import SaveProperty from "../../components/SaveProperty";

describe("SaveProperty", () => {
  const validProps = {
    userId: "43ref",
    _id: "5itfgnf",
    favouriteId: "845inf",
    onSaveProperty: jest.fn(),
    onRemoveProperty: jest.fn(),
  };

  test("falsey userId", () => {
    const { container } = render(
      <SaveProperty
        userId=""
        _id={validProps._id}
        favouriteId={validProps.favouriteId}
        onSaveProperty={validProps.onSaveProperty}
        onRemoveProperty={validProps.onRemoveProperty}
      />
    );

    expect(container).toBeEmptyDOMElement();
  });

  describe("truthy favouriteId", () => {
    test("snapshot", () => {
      const { asFragment } = render(
        <SaveProperty
          userId={validProps.userId}
          _id={validProps._id}
          favouriteId={validProps.favouriteId}
          onSaveProperty={validProps.onSaveProperty}
          onRemoveProperty={validProps.onRemoveProperty}
        />
      );

      expect(asFragment()).toMatchSnapshot();
    });

    test("renders correctly", () => {
      render(
        <SaveProperty
          userId={validProps.userId}
          _id={validProps._id}
          favouriteId={validProps.favouriteId}
          onSaveProperty={validProps.onSaveProperty}
          onRemoveProperty={validProps.onRemoveProperty}
        />
      );

      expect(screen.getByText(/saved/i)).toBeInTheDocument();
      expect(screen.queryByText(/^save$/i)).not.toBeInTheDocument();
    });
  });

  describe("falsey favouriteId", () => {
    test("snapshot", () => {
      const { asFragment } = render(
        <SaveProperty
          userId={validProps.userId}
          _id={validProps._id}
          favouriteId=""
          onSaveProperty={validProps.onSaveProperty}
          onRemoveProperty={validProps.onRemoveProperty}
        />
      );

      expect(asFragment()).toMatchSnapshot();
    });

    test("renders correctly", () => {
      render(
        <SaveProperty
          userId={validProps.userId}
          _id={validProps._id}
          favouriteId=""
          onSaveProperty={validProps.onSaveProperty}
          onRemoveProperty={validProps.onRemoveProperty}
        />
      );

      expect(screen.queryByText(/saved/i)).not.toBeInTheDocument();
      expect(screen.getByText(/^save$/i)).toBeInTheDocument();
    });
  });
});
