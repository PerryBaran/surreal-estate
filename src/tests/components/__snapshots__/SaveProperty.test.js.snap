// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SaveProperty falsey favouriteId snapshot 1`] = `
<DocumentFragment>
  <button
    class="save-property"
    type="button"
  >
    Save
  </button>
</DocumentFragment>
`;

exports[`SaveProperty truthy favouriteId snapshot 1`] = `
<DocumentFragment>
  <button
    class="remove-property"
    type="button"
  >
    <span>
      Saved
    </span>
  </button>
</DocumentFragment>
`;
